db/offchaindb/db.go

```diff
@@ -0,0 +1,483 @@
+	// with a status of:
+	//   * PropStatusPublic
+	//   * PropStatusAbandoned
+	VettedState
+	UnknownState
+)
+
+func (f ProposalStateType) String() string {
+	switch f {
+	case InvalidState:
+		return "invalid"
+	case UnvettedState:
+		return "unvetted"
+	case VettedState:
+		return "vetted"
+	default:
+		return "unknown"
+	}
+}
+
+// ProposalStateFromStr converts the string into ProposalStateType value.
+func ProposalStateFromStr(val string) ProposalStateType {
+	switch val {
+	case "invalid":
+		return InvalidState
+	case "unvetted":
+		return UnvettedState
+	case "vetted":
+		return VettedState
+	default:
+		return UnknownState
+	}
+}
+
+// ProposalDB defines the common data needed to query the proposals db.
+type ProposalDB struct {
+	dbP          *storm.DB
+	client       *http.Client
+	NumProposals int
+	_APIURLpath  string
+}
+
+// errDef defines the default error returned if the proposals db was initialized
+// correctly.
+var errDef = fmt.Errorf("ProposalDB was not initialized correctly")
+
+// NewProposalsDB opens an exiting database or creates a new db instance with the
+// provided file name. returns an initialized instance of Proposal DB, http client
+// and the politeia API URL path to be used.
+func NewProposalsDB(politeiaURL, dbPath string) (*ProposalDB, error) {
+	_, err := os.Stat(dbPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	if politeiaURL == "" {
+		return nil, fmt.Errorf("missing politeia API URL")
+	}
+
+	if dbPath == "" {
+		return nil, fmt.Errorf("missing db path")
+	}
+
+	db, err := storm.Open(dbPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// returns a http client used to query the API endpoints.
+	c := &http.Client{Transport: &http.Transport{
+		MaxIdleConns:       10,
+		IdleConnTimeout:    5 * time.Second,
+		DisableCompression: false,
+	}}
+
+	proposalDB := &ProposalDB{
+		dbP:         db,
+		client:      c,
+		_APIURLpath: politeiaURL,
+	}
+
+	return proposalDB, nil
+}
+
+// countProperties fetches the proposal count and appends it the ProposalDB instance
+// provided.
+func (db *ProposalDB) countProperties() error {
+	count, err := db.dbP.Count(&ProposalInfo{})
+	if err != nil {
+		return err
+	}
+
+	db.NumProposals = count
+	return nil
+}
+
+// Close closes the proposal DB instance created passed if it not nil.
+func (db *ProposalDB) Close() error {
+	if db == nil || db.dbP == nil {
+		return nil
+	}
+
+	return db.dbP.Close()
+}
+
+// handleGetRequests constructs the full URL path, querys the API endpoints and
+// returns the queried data in form of byte array and an error if it exists.
+func (db *ProposalDB) handleGetRequests(path, params string) ([]byte, error) {
+	response, err := db.client.Get(db._APIURLpath + path + params)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	return ioutil.ReadAll(response.Body)
+}
+
+// saveProposals adds the proposals data to the db.
+func (db *ProposalDB) saveProposals(URLParams string) (int, error) {
```

Comment:
This muddles Pi client functionality and DB functionality.  When this package is split up, it will force us to keep functions will be kept smaller, more testable, and more logically organize.